### ISSUE TITLE
全体スケジュール

### ISSUE THREAD

**Issue by flipslidersand** [flipslidersand-labs/forge#1]
## プロジェクト概要

PyTorch で書いた演算（RMSNorm / Softmax 等）に対して、Triton カーネルの実装方式とパラメータを自動探索し、
正確性・測定ノイズ・環境差を考慮した上で最速実装をキャッシュ再利用するライブラリ。

### Phase 1: IR とキャッシュ基盤（Week 1）

目標: `KernelSpec → CacheKey → SQLite` の往復が動く。GPU 不要。

- [x] `TensorSpec` / `KernelSpec` dataclass
- [x] `hash_constants` / `hash_kernel_spec`
- [x] `CacheKey.from_spec_and_env()`
- [x] `KernelRepository` (get / put / WAL)
- [x] `tests/test_ir.py` 全通過 (12 tests)
- [x] `tests/test_cache.py` 全通過 (7 tests)

**完成条件**: `pytest tests/test_ir.py tests/test_cache.py` → 19 passed ✅

---

### Phase 2: コード生成とベンチマーク（Week 2）

目標: RMSNorm の Triton コードを生成してベンチマークが動く。GPU 必須。

- [x] `codegen/triton_codegen.py` — `generate(spec, params) -> str`（Jinja2→standalone .py）
- [x] `codegen/templates/rmsnorm.py.jinja`
- [x] `benchmark/timer.py` — warmup 25 + measure 200
- [x] `benchmark/statistics.py` — `is_improvement()` (p80 < p20/1.03)
- [x] `runtime/worker.py` + `_worker_entry.py` — subprocess 隔離・タイムアウト
- [x] ~~`runtime/launcher.py`~~ → worker.py に統合
- [x] `tests/test_codegen.py`
- [x] `tests/test_benchmark.py`

**完成条件**: 生成した Triton コードが GPU で実行でき、BenchmarkResult が返る

---

### Phase 3: 探索とバリデーション（Week 3）

目標: 50 候補を探索し、PyTorch Eager 比 +20% 以上の実装をキャッシュできる。

- [x] `search/space.py` — `SearchSpace.enumerate()`（block>=N, Pascal で stages=[1]）
- [x] `search/grid.py` — `GridSearch.generate()`
- [x] `validation/test_cases.py` — 7 エッジケース（single_row/zeros/large/small/weight_randn...）
- [x] `validation/tolerance.py` — `Tolerance`（rmsnorm atol=2e-3, #4 反映）+ worker 内で検証
- [x] `orchestrator.py` — パイプライン統括
- [x] `examples/rmsnorm_search.py` — end-to-end デモ
- [x] `tests/test_search.py` / `test_validation.py` / `test_orchestrator.py`

**完成条件**: `python examples/rmsnorm_search.py` が動き、speedup を報告できる

---

### Phase 4: 複数バリアント（Week 4〜5）

目標: 実装方式（single_row / multi_row / two_pass / fused_weight）を探索できる。

- [x] `codegen/templates/rmsnorm_multi_row.py.jinja`
- [x] `codegen/templates/rmsnorm_two_pass.py.jinja`
- [x] `search/space.py` に `variant` 軸を追加（variant-aware enumerate）
- [x] `search/random_search.py`（決定的 LCG シャッフル）

**完成条件**: `--variants all` で複数バリアントを含む探索が動く。`@triton.autotune` 相当以上の性能

---

### Phase 5: LLM 候補生成（Week 6+）

目標: LLM が構造化命令を出し、探索候補を拡張できる。

- [x] `search/candidate.py` — `CandidateGenerator` Protocol + `HistoryEntry`
- [x] `search/llm_generator.py` — Claude(messages.parse)で構造化候補生成、注入可能
- [x] Grid/Random を Protocol に統一

**完成条件**: LLM 生成候補が既存パイプラインに流れ、検証・ベンチマークが通る

---

### Phase 6: `@forge.optimize` デコレータ（Week 7+）

目標: デコレータ 1 つで end-to-end が動く。2 回目以降はキャッシュのみ。

- [x] `decorator.py` — `optimize()` デコレータ（shape毎に探索→in-processキャッシュ）
- [x] `lowering/registry.py` — `torch.fx` call_function 多重集合でマッチ
- [x] `lowering/rmsnorm.py` + `lowering/__init__.py identify()` — RMSNorm 判定
- [x] `runtime/loader.py` — worker/decorator 共通の kernel ローダ

**完成条件**:
```python
@forge.optimize(budget=50)
def rmsnorm(x, weight, eps=1e-6): ...
```
初回 → 探索・キャッシュ、2 回目以降 → 探索ゼロで実行

---

## 依存関係・順序の制約

```
Phase 1 → Phase 2 → Phase 3 → Phase 4
                                  ↓
                              Phase 5 → Phase 6
```

Phase 1 が全フェーズの土台。Phase 2 は GPU が必要なため環境セットアップ確認が先決。

## リスク・難所

| リスク | 対策 |
|---|---|
| GPU 測定ノイズで誤判定 | p80 < p20/1.03 の統計判定（ADR-004） |
| CUDA エラーで親プロセスが死ぬ | subprocess 使い捨てワーカー（ADR-003） |
| キャッシュキー衝突 | Triton/CUDA/torch バージョンを全部含める（ADR-002） |
| RMSNorm の縮約誤差 | `atol=1e-3, rtol=1e-2` で許容 |
| LLM がコード破壊 | 構造化 JSON のみ受け取る（ADR-005） |

**Comment by flipslidersand** [comment]
## Phase 4 完了 (commit ad856c9)

コード構造の探索が可能に。

- **variants**: single_row / multi_row（ROWS行/program）/ two_pass（タイル化縮約、BLOCK<N 可）
- codegen を `(op_type, variant)` でディスパッチ、`SearchParams.rows_per_program` 追加
- `SearchSpace` を variant-aware に（two_pass はタイル<=N、multi_row は rows>1）
- `RandomSearch`（variant でグリッドが膨らむため決定的サンプリング）

### GTX 1080 実走（512×4096）
```
3 variant 全てコンパイル・正確（max_diff 1.95e-3）
60 候補 → 42 correct（single_row/multi_row の fp16 acc を棄却、two_pass fp16 は fp32縮約で通過）
best: single_row 62.4us vs F.rms_norm 390us = 6.25x
two_pass は N=4096 では遅い（70-79us）= 大きな N 向け構造であることを実測で確認
```
テスト 77 passed（73 CPU + 4 GPU）

### 次（Phase 5 / 別途 Issue 化候補）
- baseline に `torch.compile` / `@triton.autotune` 追加（per-candidate でなく one-shot 計測に分離）
- CandidateGenerator Protocol → LLM 候補生成

**Comment by flipslidersand** [comment]
## Phase 5 完了 (commit b81303a)

LLM を探索器の一実装として差し込み可能に。

- `CandidateGenerator` Protocol に Grid/Random/LLM を統一（`HistoryEntry` でフィードバック）
- `LLMGenerator`: Claude (`claude-opus-4-8`) を `messages.parse` + Pydantic スキーマで呼び、**自由コードでなく構造化パラメータ＋仮説のみ**を出力（ADR-005）
- 不正候補（block 不整合 / fp16-acc / variant-spec ミスマッチ）は coerce して棄却・dedup・budget cap
- **API 呼び出しは注入可能**（`propose_fn`）、`anthropic`/`pydantic` は optional `[llm]` extra + 遅延 import → API キー無し・anthropic 未インストールでも全テスト通過

### 検証（GTX 1080、propose_fn 注入）
LLM 形式の候補が codegen→worker→検証→ベンチ→キャッシュを通過。fp16-acc 提案は INCORRECT で棄却、best single_row fp32 を 7.4x でキャッシュ。

テスト 94 passed（90 CPU + 4 GPU、うち新規 17 は完全オフライン）。

### 残（Phase 6 / 任意）
- `@forge.optimize` デコレータ + torch.fx グラフ→KernelSpec（lowering）
- ライブ LLM 探索（要 `ANTHROPIC_API_KEY`）と grid/random との効果比較
- 反復ループ（history を渡して LLM が前ラウンドから学習）

**Comment by flipslidersand** [comment]
## 🎉 Phase 6 完了 — 全フェーズ達成 (commit 9d3336d)

`@forge.optimize` デコレータ完成。プレーンな PyTorch 関数 1 つ → 自動最適化。

```python
@forge.optimize(budget=12)
def rmsnorm(x, weight, eps=1e-6):
    return x * torch.rsqrt(torch.mean(x * x, dim=-1, keepdim=True) + eps) * weight
```

- `lowering/`: torch.fx で trace → call_function 名の多重集合でパターンマッチ（rmsnorm = `{mul:3, mean:1, add:1, rsqrt:1}`）。未知/trace不能は None → eager フォールバック（コンパイラ沼を回避）
- `decorator.py`: 呼び出し引数を tensor=入力 / scalar=定数 に分離。shape 毎に初回探索 → 勝者カーネルを in-process ロード、同 shape は再利用、新 shape は再探索（SQLite ヒット）
- `runtime/loader.py`: worker(subprocess) と decorator(in-process) 共通の kernel ローダに DRY 化

### GTX 1080 実走（decorator_demo.py）
```
初回: 12候補探索(37s) → 正確な最速を確定・キャッシュ、出力は eager と一致
2回目: in-process キャッシュ → 0.28ms（カーネル起動のみ）
```

テスト **106 passed**（94 CPU + 12 GPU、新規 8 lowering + 4 decorator）。

---

## プロジェクト総括

Phase 1〜6 完了。RMSNorm を対象に、総評で指摘された設計の落とし穴をすべて踏まえて実装:
- ✅ IR(KernelSpec) 中心の再現可能パイプライン（LLM中心にしない）
- ✅ 環境込み CacheKey（torch/triton/cuda/cc）
- ✅ 統計的採用判定（p80<p20/1.03、ノイズ実測0.7%）
- ✅ 演算ごとの許容誤差（atol=2e-3、実測根拠）+ 7エッジケース
- ✅ subprocess 隔離（CUDAエラーで親が死なない）
- ✅ パラメータ＋コード構造(variant)の探索
- ✅ LLM は差し替え可能な候補生成器の一つ（構造化出力のみ）
- ✅ `@forge.optimize` は UI 層に徹し内部複雑性を隠蔽

### 今後の発展余地（任意）
- Softmax 等への op 拡張、torch.compile/autotune baseline、探索コスト考慮の採用判定、ライブ LLM 反復探索

**Comment by flipslidersand** [comment]
## 発展: Softmax 追加 — アーキの汎用性を証明 (commit f20cced)

同じ `@forge.optimize` が **Softmax も最適化**。全レイヤーを2演算目で通し、RMSNorm 専用でないことを実証。

- **codegen**: softmax single_row + multi_row テンプレート（safe softmax: 行max減算→fp32でexp/sum、マスク外は-infでsumに不寄与）
- **reference**: softmax ground truth(fp32) + F.softmax baseline
- **lowering**: `{softmax:1}` fxパターン登録。`torch.softmax(x, dim=-1)`（定数dim）を認識。dim を引数で渡す形は trace 不能→eager フォールバック
- **validation**: test_cases を op_type で分岐、softmax 専用エッジケース（大きい値で safe-softmax のオーバーフロー耐性を検証）、atol=2e-3
- **orchestrator**: codegen を try/except で囲み、テンプレート無し variant（softmax two_pass）を候補ごとスキップ（クラッシュしない）

### GTX 1080 実走
```
@forge.optimize softmax → 20候補探索 → best single_row 4096/4/fp32 = 164.9us
出力は eager と一致、各行の和 ≈ 1
```

テスト **112 passed**（99 CPU + 13 GPU）。

### 設計上の収穫
op 追加に必要なのは「テンプレート + reference + lowering パターン + tolerance」の4点のみ。中核パイプライン（探索・検証・ベンチ・キャッシュ・デコレータ）は無改修で再利用できた = IR 中心設計が効いている。

### 残（任意）
torch.compile/autotune baseline、ライブ LLM 反復探索、探索コスト考慮の採用判定、さらなる op（layernorm/gelu 等）

**Comment by flipslidersand** [comment]
## ロードマップ issue を起票

- ✅ #7 Softmax 追加（完了・クローズ）
- 📋 #8 op 拡張: LayerNorm / GELU
- 📋 #9 baseline 拡張: torch.compile / @triton.autotune
- 📋 #10 ライブ LLM 反復探索
- 📋 #11 探索コスト考慮の採用判定（独自性テーマ）

**Comment by flipslidersand** [comment]
Phase 1〜6 全完了。全チェックボックス達成。
